### PR TITLE
GH-1028 Adds MegaChart widget

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes GH-1028
Fixes GH-1027

http://rawgit.com/jbrundage/Visualization/MEGA_CHART_CLUSTER/demos/dermatology.html?src/cluster/MegaChart

Signed-off-by: Jay Brundage <jaman.brundage@lexisnexis.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
1. The X/Y labels should have the same font options that the Surface Title has (bold, size etc...)
2. Rather than X/Y Axis it should be "Domain" and "Value" Axis (the PP label) - I will be changing the XYAxisWidget at some point - mainly as it is very confusing when the orientation is changed?
3.  The legend shrink wrap won't work due to the Table auto resizing to consume all the space (open a new issue to add "fixedSize" PP to the Table).
4.  Rename Cluster folder to "Composite"
5.  The Composite folder will need to be added to the build and to the non AMD demo/test pages.

**Comment by GordonSmith** [comment]
6.  The mega chart should probably not show any of the borders if a 1D or 2D chart is selected (or maybe they are not options for the Mega Chart?

File: src/other/Legend.js [hunk 1]
```diff
@@ -95,7 +95,9 @@
         
         var table = element.select(".tableDiv > table");
         var tableRect = table.node().getBoundingClientRect();
-        var elementRect = element.node().getBoundingClientRect();
+        var elementRect = this._parentElement.node().getBoundingClientRect();
+        console.log("elementRect.height:");
+        console.log(elementRect.height);
```
Comment (GordonSmith) [inline src/other/Legend.js]: remove logging

File: src/common/Text.js [hunk 2]
```diff
@@ -60,10 +62,15 @@
                 break;
         }
         var yOffset = -(bbox.y + bbox.height / 2);
-
+        
+        if (Math.abs(context._rotation) === 90) {
+            xOffset = -(bbox.y + bbox.height / 2);
+            yOffset = 0;
+        }
```
Comment (GordonSmith) [inline src/common/Text.js]: Why is this needed for +/-90 and not for say +/-89?

File: src/other/Table.js [hunk 3]
```diff
@@ -134,7 +136,13 @@
     Table.prototype.update = function (domNode, element) {
         HTMLWidget.prototype.update.apply(this, arguments);
         var context = this;
-       
+        if(this.fixedSize()){
+            var box = d3.select(".tableDiv > table").node().getBoundingClientRect();
+            element.style({width: box.width + "px",height: box.height + "px"});
```
Comment (GordonSmith) [inline src/other/Table.js]: Would simply removing the width + height styles here be simpler?

File: src/composite/MegaChart.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    MegaChart.prototype = Object.create(Surface.prototype);
+    MegaChart.prototype.constructor = MegaChart;
+    MegaChart.prototype._class += " composite_MegaChart";
+
+    MegaChart.prototype.publish("xAxisTitle","","string","Title to display for X-Axis", null, {tags:["Basic"]});
+    MegaChart.prototype.publish("yAxisTitle","","string","Title to display for Y-Axis", null, {tags:["Basic"]});
+    
+    MegaChart.prototype.publish("showLegend",false,"boolean","Show/Hide Legend", null, {tags:["Basic"]});
+    
+    MegaChart.prototype.publish("gutter", 8, "number", "Gap Between Widgets",null,{tags:["Basic"]});
```
Comment (GordonSmith) [inline src/composite/MegaChart.js]: "Gap Between Cells"

**Comment by GordonSmith** [comment]
Looks good - some small comments/questions above.

File: src/composite/MegaChart.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    if (typeof define === "function" && define.amd) {
+        define(["../layout/Surface", "../layout/Border", "../common/Text", "../other/Legend"], factory);
+    } else {
+        root.composite_MegaChart = factory(
+                root.layout_Surface,
+                root.layout_Border,
+                root.common_Text,
+                root.other_Legend
+            )
+        ;
```
Comment (GordonSmith) [inline src/composite/MegaChart.js]: Should be one line for consistency

File: src/composite/MegaChart.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    }
+}(this, function (Surface, Border, Text, Legend) {
+    function MegaChart() {
+        Surface.call(this);
+
+        this._tag = "div";
```
Comment (GordonSmith) [inline src/composite/MegaChart.js]: Not needed

File: src/composite/MegaChart.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    
+    MegaChart.prototype.publish("gutter", 8, "number", "Gap Between Widgets",null,{tags:["Basic"]});
+    
+    MegaChart.prototype.publish("chart",null,"widget","Chart widget to display in 'center' of MegaChart", null, {tags:["Basic"]});
+
+    MegaChart.prototype.enter = function (domNode, element) {
+        this._layout = new Border();
+        
+        this._layout.setContent("center",this.chart());
+        if(this.showLegend()){
```
Comment (GordonSmith) [inline src/composite/MegaChart.js]: Needs to be in update

File: src/composite/MegaChart.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            this._layout.setContent("right", new Legend().targetWidget(this.chart()));
+        }
+        if(this.xAxisTitle()){
```
Comment (GordonSmith) [inline src/composite/MegaChart.js]: Needs to be in update

File: src/composite/MegaChart.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            var xAxisTitleText = new Text().text(this.xAxisTitle());
+            this._layout.setContent("bottom", xAxisTitleText);
+        }
+        if(this.yAxisTitle()){
```
Comment (GordonSmith) [inline src/composite/MegaChart.js]: Needs to be in update

**Comment by GordonSmith** [comment]
One more comment:
I think I would create the X/YAxis title widgets and the legend widget in the enter method and then just add/remove them in the update method if showLegend === true and if the XYAxis text !== "" 
As it is, there will be issues when the marshaller starts to use them (and with Persist).

Also if you do it that way, you can add proxyPublish calls for X/YAxis titles.

File: src/common/Text.js (hunk 2 above)
Comment (jbrundage) [inline src/common/Text.js]: You're right... in its current state this offset needs some sort of trigonometric equation to determine its xOffset and yOffset... but that assumes that its parent isn't rotated? I'm not sure... maybe this implementation needs to be refactored?

Comment (mlzummo) [inline src/common/Text.js]: XY Axis label rotation math should be applied here ?

Comment (jbrundage) [inline src/common/Text.js]: Similar math, but both x and y taken into consideration this time

File: src/common/SVGWidget.js [hunk 10]
```diff
@@ -131,7 +137,7 @@
     SVGWidget.prototype.postUpdate = function (domeNode, element) {
         Widget.prototype.postUpdate.apply(this, arguments);
         if (this._drawStartPos === "origin" && this._target instanceof SVGElement) {
-            this._element.attr("transform", "translate(" + (this._pos.x - this._size.width / 2) + "," + (this._pos.y - this._size.height / 2) + ")scale(" + this._scale + ")");
+            this._element.attr("transform", "translate(" + (this._pos.x - this._size.width / 2) + "," + (this._pos.y - this._size.height / 2) + ")");
```
Comment (GordonSmith) [inline src/common/SVGWidget.js]: Why are you removing Scale from here?

File: src/common/SVGWidget.js [hunk 11]
```diff
@@ -120,6 +120,12 @@
         return this;
     };
 
+    SVGWidget.prototype.rotation = function (_) {
+        if (!arguments.length) return this._rotation;
+        this._rotation = _;
+        return this;
+    };
+
```
Comment (GordonSmith) [inline src/common/SVGWidget.js]: If rotation is not applied within Widget then the setter/getter should not be declared here (IOW it should be in Text.js)

File: src/common/Text.js [hunk 12]
```diff
@@ -59,11 +62,18 @@
...
+        var r = xShift + yShift;
+        var theta = -context._rotation * Math.PI/180;
+        
+        xOffset = Math.round(r * Math.sin(theta));
+        yOffset = Math.round(r * Math.cos(theta));
+        
         this._textElement
             .style("text-anchor", this.anchor())
-            .attr("transform", function (d) { return "translate(" + xOffset + "," + yOffset + ")"; })
+            .attr("transform", function (d) { return "translate(" + xOffset + "," + yOffset + ")scale(" + context._scale + ")rotate(" + context._rotation + ")"; })
```
Comment (GordonSmith) [inline src/common/Text.js]: I am not sure Scale should be here.

File: src/other/Table.js (hunk 3 above)
Comment (jbrundage) [inline src/other/Table.js]: I don't follow what you're suggesting... removing the width + height styles of this.element()?

The goal of lines 140 and 141 are to grab the width and height of the <table> and apply them to the target's <div>.

**Comment by GordonSmith** [comment]
Looks good - @jbrundage Will you open a new issue to add button(s) to the Surface title to show/hide the legend?

**Comment by GordonSmith** [comment]
@jbrundage Also open an issue to add composite to the builds + non amd test pages.
